core/camera/src/main/java/com/google/jetpackcamera/core/camera/FeatureGroupData.kt

```diff
@@ -0,0 +1,164 @@
+
+    /** Feature value that is not usable with CameraX feature groups API. */
+    data class Ungroupable<T>(val featureValue: T) : FeatureGroupData<T>
+}
+
+/**
+ * Returns whether a collection of [FeatureGroupData] represents that CameraX feature group
+ * query API should be used.
+ *
+ * This is based on the fact that feature group query API should be used whenever there's at least
+ * two [GroupableFeature].
+ */
+fun Collection<FeatureGroupData<*>>.requiresFeatureGroupQuery(): Boolean {
+    return filter { it is ExplicitlyGroupable }.size >= 2
+}
+
+/**
+ * Returns whether a collection of [FeatureGroupData] is invalid.
+ *
+ * The collection should not be used for camera session i.e. invalid whenever it has a
+ * [Ungroupable] element and [requiresFeatureGroupQuery] is true for the collection.
+ */
+internal fun Collection<FeatureGroupData<*>>.isInvalid(): Boolean {
+    return requiresFeatureGroupQuery() && any { it is Ungroupable }
+}
+
+/**
+ * Converts this [DynamicRange] to a [FeatureGroupData].
+ *
+ * This allows the dynamic range to be used in CameraX feature group compatibility checks.
+ */
+fun DynamicRange.toFeatureGroupData(): FeatureGroupData<DynamicRange> {
+    return when (this) {
+        DynamicRange.SDR -> InexplicitlyGroupable
+        DynamicRange.HLG10 -> ExplicitlyGroupable(GroupableFeature.HDR_HLG10)
+    }
+}
+
+/**
+ * Converts this [VideoQuality] to a [GroupableFeatureData].
+ *
+ * This allows the video quality to be used in CameraX feature group compatibility checks.
+ */
+fun VideoQuality.toFeatureGroupData(): FeatureGroupData<VideoQuality> {
+    return when (this) {
+        SD -> ExplicitlyGroupable(GroupableFeatures.SD_RECORDING)
+        HD -> ExplicitlyGroupable(GroupableFeatures.HD_RECORDING)
+        FHD -> ExplicitlyGroupable(GroupableFeatures.FHD_RECORDING)
+        UHD -> ExplicitlyGroupable(GroupableFeatures.UHD_RECORDING)
+        UNSPECIFIED -> InexplicitlyGroupable
+    }
+}
+
+/**
+ * Converts this [ImageOutputFormat] to a [GroupableFeatureData].
+ *
+ * This allows the image output format to be used in CameraX feature group compatibility checks.
+ */
+fun ImageOutputFormat.toFeatureGroupData(): FeatureGroupData<ImageOutputFormat> {
+    return when (this) {
+        JPEG -> InexplicitlyGroupable
+        JPEG_ULTRA_HDR -> ExplicitlyGroupable(GroupableFeature.IMAGE_ULTRA_HDR)
+    }
+}
+
+/**
+ * Converts this [StabilizationMode] to a [GroupableFeatureData].
+ *
+ * This allows the stabilization mode to be used in CameraX feature group compatibility checks.
+ */
+fun StabilizationMode.toFeatureGroupData(): FeatureGroupData<StabilizationMode> {
+    return when (this) {
+        OFF -> InexplicitlyGroupable
+        ON -> ExplicitlyGroupable(GroupableFeature.PREVIEW_STABILIZATION)
+        HIGH_QUALITY -> ExplicitlyGroupable(GroupableFeatures.VIDEO_STABILIZATION)
+        AUTO, // AUTO should be resolved to concrete value before this API is called
```

Comment:
Throw an `IllegalStateException` rather than returning `Ungroupable` for `AUTO`. It's a programming error to try to call `toFeatureGroupData` without first resolving it to a concrete stabilization mode.